src/server/tiering/external_alloc.cc

```diff
@@ -478,12 +496,23 @@ void ExternalAllocator::FreePage(Page* page, SegmentDescr* owner, size_t block_s
   BinIdx bidx = ToBinIdx(block_size);
 
   // Remove fast allocation reference.
-  if (free_pages_[bidx] == page) {
-    free_pages_[bidx] = &empty_page;
+  if (page->free_chained) {
```

Comment:
According to your change any page at this point must be part of `free_pages_` list. Moreover, seems reasonable to add this invariant to class Page description:
```
Page can be exactly in either these 3 states:
1. unitialized - with no blocks being allocated - segment_inuse will be 0 in that case, 1 otherwise.
2. Partly utilized by 1 or more blocks, with available > 0 in that case. It must be present in free_pages_ list then.
3. Fully utilized, with available==0, in that case it's not part of free_pages_ list.
```
If what I wrote is correct, you do not need `free_chained` variable.

Reply:
Looks like it, I just made sure it's versatile and won't break. However worst we do with this is just do a pointless search, so nothing critical